[native assets] Graduate to preview (#6612)
dart-lang/site-www: src/content/interop/c-interop.md

To have Dart create FFI wrappers from C header files,
use the [`package:ffigen`][ffigen] binding generator.

> ## Dart packages containing native code {:#native-assets}

Reviewer: Let's change the ID of this fragment, but add an anchor so the old one keeps working.

```suggestion
<a id="native-assets" aria-hidden="true"></a>

## Dart packages containing native code {: #build-hooks }
```

Author: Should this be area-hidden ?

Reviewer: We tend to, as it doesn't have any semantic meaning. It's just an invisible target for the browser to navigate to.

Author: I mean "area-hidden", but apparently it is "aria-hidden" and I don't know HTML nowadays anymore.

Reviewer 2: I notice that this change breaks the 'How to X' flow of headings on this page. Let's make sure to keep that style. I would either revert this change or update this heading to say something like:

`Work with native code`

or

`Work with Dart packages that contain native code`

I don't think we should hide this heading from the right-pane TOC.

Reviewer: > I notice that this change breaks the 'How to X' flow of headings on this page. Let's make sure to keep that style.

That makes sense, I didn't consider that. I opened https://github.com/dart-lang/site-www/pull/6614 with an updated title.

> As @parlough mentioned, we should not change that that old slug (`#native-assets`}
> 
> We shouldn't need `<a id="native-assets" aria-hidden="true"></a>` if you use the original slug.

I'd prefer having the updated slug (`build hooks`) and leaving `native-assets` as a migratory fragment as it's how the feature is going to be referred to.

:::note

> The dart build hooks support is in **preview** and can be used in Dart & Flutter main channel.

Reviewer: ```suggestion
Support for build hooks is in **preview** and
can be used on the `main` channel.
```

:::

> The _Build Hooks_ (formerly known as _Native Assets_) feature enables Dart
> packages to contain more than just Dart source code. Dart packages can now
> contain native code assets that will be transparently built, bundled and made
> available at runtime.

Reviewer: ```suggestion
Dart _build hooks_ (formerly known as _native assets_) enable
packages to contain more than just Dart source code.
Packages can now contain native code assets that are
transparently built, bundled, and made available at runtime.
```

This feature simplifies how Dart packages depend on and use native code:

> * Build the native code or obtain the binaries using a package's build hook
> (`hook/build.dart`).

Reviewer: ```suggestion
* Build the native code or obtain the binaries using a
  package's build hook in `hook/build.dart`.
```

> * Dart and Flutter bundle the native code [`Asset`][]s that the build hook

Reviewer 3: the native code `Asset` -> the `CodeAsset`s?

  reports.
* Access the code assets at runtime through declarative
  `@Native<>() extern` functions using the [`assetId`][].

> Flutter and standalone Dart automatically bundle the native code of all packages
> used by the app and makes them available at runtime. This works for `flutter
> (run|build)` as well as `dart (run|build)` .

Reviewer: ```suggestion
Flutter and standalone Dart automatically bundle the
native code of all packages used by the app and make them available at runtime.
This works for `flutter (run|build)` as well as `dart (run|build)`.
```

Author: make it ? (instead of make them)

Reviewer: Good point! :)

```suggestion
Flutter and standalone Dart automatically bundle the
native code of all packages used by the app and make it available at runtime.
This works for `flutter (run|build)` as well as `dart (run|build)`.
```

To have Dart create FFI wrappers from C header files,
use the [`package:ffigen`][ffigen] binding generator.

## Dart packages containing native code {:#native-assets}

:::note
Support for build hooks is in **preview** and
can be used on the `main` channel.
:::

> Dart _build hooks_ (formerly known as _native assets_) enable

Reviewer 2: I am worried about us switching terminology here. It makes the sections that follow not make as much sense. If this change was a personal preference, let's stick with the original terminology.

If this change is tied to code changes that update the terminology, you should also look at the subsections and update them as needed to match.

## Build and bundle native assets {:#native-assets}

:::note
The dart build hooks support is in **preview**.
:::

The _Build Hooks_ (formerly known as _Native Assets_) feature should
resolve a number of issues associated with
the distribution of Dart packages that depend on native code.
It does so by providing uniform hooks for integrating with various
build systems involved in building Flutter and standalone Dart applications.

This feature should simplify how Dart packages depend on and use native code.
Build hooks should provide the following benefits:

* Build the native code or obtains the binaries
  using a package's `hook/build.dart` build hook.
* Bundle the native [`Asset`][] that the `build.dart` build hook reports.
* Make native code assets available at runtime through
  declarative `@Native<>() extern` functions using the [`assetId`][].

> When you [opt in](#opt-in-to-the-experiment) to the native experiment,

Author: My understanding of preview is that is not yet available in stable (you can't flip the experiment to true in stable), but it is available in the other channels and on by default.

## Build and bundle native assets {:#native-assets}

:::note

> The native assets are in **preview**,

Reviewer 3: Could we rephrase this as `The dart build hooks support is in preview` and also replace most other "native assets" with "dart hooks" / "dart build hooks"?
